[Index management] Project level retention support

sabarasaba opened the pull request:
Fixes: https://github.com/elastic/kibana/issues/193160

## Summary

This PR adds support for project level retention in index management data streams UI. At the moment **this enforcement is only** in place for stateful security projects.

##### How to test
* Start ES with `node scripts/es serverless --tag latest --projectType security -E data_streams.lifecycle.retention.factory_default=200d -E data_streams.lifecycle.retention.factory_max=200d -E data_streams.lifecycle.retention.max=200d` and kibana with `yarn serverless-security`
* Navigate to console and create a data stream following [this instructions](https://www.elastic.co/guide/en/elasticsearch/reference/current/set-up-a-data-stream.html#create-component-templates)
* Navigate to `Index Management` -> `Data streams`
* Verify that:
  * Changes are only applied to security solution
  * A callout is shown above the data streams page with a link to cloud (you might need to configure cloud url locally in order to have it working) 
  * The callout is dismissable and the setting for it is stored in localstorage.
  * When editing the data retention, there should be an enforced limit in the UI based on the global retention returned by the lifecycle api.
  
  
```
# Configure cloud locally by adding the following to config/kibana.dev.yml and restarting kibana
xpack.cloud.id: 'test'
xpack.cloud.base_url: "https://cloud.elastic.co"
xpack.cloud.deployment_url: "/deployments/deploymentId"
```


<details>
<summary>Screenshots</summary>

![Screenshot 2024-09-24 at 16 05 49](https://github.com/user-attachments/assets/b44e6b24-0f8f-4a41-b55a-5f79fae70f48)
![Screenshot 2024-09-24 at 16 05 54](https://github.com/user-attachments/assets/240b6c53-16d0-48d2-a86c-58c7e2e16b4d)
![Screenshot 2024-09-30 at 15 09 49](https://github.com/user-attachments/assets/b3ea2285-50fc-4147-846c-f60994057b1e)




</details>

<!--ONMERGE {"backportTargets":["8.16","8.x"]} ONMERGE-->

<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!

ElenaStoeva reviewed (commented):
Thanks for working on this @sabarasaba! I tested locally and have a few questions:

- The "Get started" button in the callout doesn't do anything, I guess because I haven't configured the cloud url locally. Could you add instructions on how to do this?

- The edit data retention modal in this PR looks a bit different from the design in the [issue](https://github.com/elastic/kibana/issues/193160) that is related to this PR. For example, the description is missing, the units are not fixed to "days" only, and the toggle description is different. Is the design in the issue outdated?


![image](https://github.com/user-attachments/assets/b1ef893f-7a85-46db-9e63-012b2027cb1a)

<img width="487" alt="Screenshot 2024-09-30 at 13 37 08" src="https://github.com/user-attachments/assets/3eca03e4-c76d-4e64-bd89-d1b89e526e4a">

sabarasaba commented:
Thanks for having a look @ElenaStoeva!

> * The "Get started" button in the callout doesn't do anything, I guess because I haven't configured the cloud url locally. Could you add instructions on how to do this?

Added a small comment in the PR description about this.

> * The edit data retention modal in this PR looks a bit different from the design in the [issue](https://github.com/elastic/kibana/issues/193160) that is related to this PR. For example, the description is missing, the units are not fixed to "days" only, and the toggle description is different. Is the design in the issue outdated?

For having days as a fixed unit, there was quite some UX complexity hidden behind that for making this retrocompatible so we decided with @jovana-andjelkovic and @debadair to leave it as is for now. About the missing title, I've updated the copy (https://github.com/elastic/kibana/pull/193715/commits/dc6b24499932ec075655d1e92a411c324afc157e) thanks for noticing!

ElenaStoeva commented:
Thanks for updating the modal copy and for adding the cloud URL instructions! I tested again but realized that when you set a unit different from 'days', only the input number is compared to the maximum period, but the unit is not considered. For example, it doesn't allow 1000 minutes, even though it's lower than 200 days, just because 1000 > 200. Could we fix this somehow?

<img width="466" alt="Screenshot 2024-09-30 at 15 46 42" src="https://github.com/user-attachments/assets/31de4c45-4756-4b63-9682-aefd34904dc7">


Also, one more thing I noticed: if the error is displayed, the Save button is still enabled. If you click it, it gives the impression that the change has been saved despite the error, but it actually wasn't saved.


https://github.com/user-attachments/assets/0adcda49-56d8-4cc5-b6cc-7844ab5da9af

sabarasaba commented:
Good catch! I've updated the comparator function with https://github.com/elastic/kibana/pull/193715/commits/ac0625639edd02576d7ba6b7813a5fec71a3ef98, this converts all units to a single unit (minutes) and does a simple value comparison then to see which one is bigger.

ElenaStoeva reviewed (commented):
Hey @sabarasaba, thanks for adding the changes! Just tested again and the comparison works better now, but I see an issue with the errors not being in sync with the values 😅 

For example, if you set 10000 minutes, and then change to "days", no error will be displayed and you can save the retention period to 10000 days. Also, if you set 1000 days (and an error is displayed) and then change to "minutes", the error is still displayed and the Save button is disabled.


https://github.com/user-attachments/assets/ecfdfe3c-4130-4160-b476-6896535826fa

sabarasaba commented:
Nice catch! We werent validating these fields properly before. I've fixed it with https://github.com/elastic/kibana/pull/193715/commits/50c0382ed8696a72969a6ec2f94d418e2266bdc4 by triggering a validation event on the retentionPeriod field whenever the form data changes.

ElenaStoeva reviewed (approved):
Thanks for fixing the issues with the validation @sabarasaba! Tested locally and I don't see any more issues. Changes also lgtm with a few nits.

sabarasaba commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/data_stream_list.tsx:
```diff
@@ -81,6 +86,10 @@ export const DataStreamList: React.FunctionComponent<RouteComponentProps<MatchPa
 
+  const [projectLevelRetentionCallout, setprojectLevelRetentionCallout] =
+    useStateWithLocalStorage<boolean>(SHOW_PROJECT_LEVEL_RETENTION, true);
+  const shouldShowProjectLevelRetention = enableProjectLevelRetentionChecks;
```
Nice catch, was a reminder of some logic that got refactored out!

sabarasaba commented on x-pack/plugins/index_management/__jest__/client_integration/home/data_streams_project_level_retention.test.ts:
```diff
@@ -0,0 +1,161 @@
+    testBed.find('filter-option-d').simulate('click');
+
+    await act(async () => {
+      testBed.find('saveButton').simulate('click');
```
Not entirely sure what was going on under the hood there, but I spent few hours on that and it was the only way I could manage to get the error to bubble up 😬

ElenaStoeva replied:
I understand...sometimes I've had similar issues with tests 😅 I was just thinking if the button gets disabled because we already clicked on it once, and not because there is an error... If that's true the test would pass even if we set the value to something lower than the maximum retention period. Would there be some way to check if there is an error message displayed instead of checking the button? 🤔

elasticmachine commented:
Pinging @elastic/kibana-management (Team:Kibana Management)

sabarasaba commented:
@elasticmachine merge upstream

jloleysens reviewed (approved):
YAML changes LGTM. @sabarasaba reach out to me if you'd like to roll out this configuration change more incrementally to serverless environments (first to QA, then staging then prod). This approach will promote this config to all envs on next rollout.

SiddharthMantri reviewed (approved):
Config changes LGTM!

sabarasaba commented:
@elasticmachine merge upstream

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/240525)
* Commit: 9aaf7bdae9ff9d26bac998e481f1c769b6c89a31

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/40471aedb9816a81e9ae6fcc143a96162068358d) | [after](https://github.com/elastic/kibana/commit/9aaf7bdae9ff9d26bac998e481f1c769b6c89a31) | diff |
| --- | --- | --- | --- |
| `indexManagement` | 692 | 693 | +1 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/40471aedb9816a81e9ae6fcc143a96162068358d) | [after](https://github.com/elastic/kibana/commit/9aaf7bdae9ff9d26bac998e481f1c769b6c89a31) | diff |
| --- | --- | --- | --- |
| `indexManagement` | 246 | 247 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/40471aedb9816a81e9ae6fcc143a96162068358d) | [after](https://github.com/elastic/kibana/commit/9aaf7bdae9ff9d26bac998e481f1c769b6c89a31) | diff |
| --- | --- | --- | --- |
| `indexManagement` | 684.4KB | 687.3KB | +2.9KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/40471aedb9816a81e9ae6fcc143a96162068358d) | [after](https://github.com/elastic/kibana/commit/9aaf7bdae9ff9d26bac998e481f1c769b6c89a31) | diff |
| --- | --- | --- | --- |
| `indexManagement` | 45.2KB | 45.2KB | +81.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/40471aedb9816a81e9ae6fcc143a96162068358d) | [after](https://github.com/elastic/kibana/commit/9aaf7bdae9ff9d26bac998e481f1c769b6c89a31) | diff |
| --- | --- | --- | --- |
| `indexManagement` | 251 | 252 | +1 |


</details>


cc @sabarasaba

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240525","number":240525,"commit":"9aaf7bdae9ff9d26bac998e481f1c769b6c89a31"}],"number":240525}
buildkite-pr-comment-->

sabarasaba commented:
@elasticmachine merge upstream

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/241513)
* Commit: b5f92dc1b8acee94df4d863154ff967a3eadbc7c

### Failed CI Steps
* [:pipeline::magic_wand: Run Dynamic Pipeline: .buildkite/scripts/pipelines/pull_request/pipeline.sh](https://buildkite.com/elastic/kibana-pull-request/builds/241513#01927b17-0a5a-4e22-ba73-b1f7690fa90f)

### History
* :green_heart: [Build #240525](https://buildkite.com/elastic/kibana-pull-request/builds/240525) succeeded 9aaf7bdae9ff9d26bac998e481f1c769b6c89a31

cc @sabarasaba

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/241513","number":241513,"commit":"b5f92dc1b8acee94df4d863154ff967a3eadbc7c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240525","number":240525,"commit":"9aaf7bdae9ff9d26bac998e481f1c769b6c89a31"}],"number":241513}
buildkite-pr-comment-->

sabarasaba commented:
@elasticmachine merge upstream

kibanamachine commented:
Starting backport for target branches: 8.16, 8.x

https://github.com/elastic/kibana/actions/runs/11464452690

kibanamachine commented:
Starting backport for target branches: 8.16, 8.x

https://github.com/elastic/kibana/actions/runs/11464452647

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.16|Backport failed because of merge conflicts|
|❌|8.x|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 193715
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.16|Backport failed because of merge conflicts|
|❌|8.x|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 193715
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/11478886410

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.x|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 193715
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
